### ISSUE TITLE
✨ Pool Candidate Status Update

### ISSUE THREAD

**Issue by petertgiles** [GCTC-NTGC/gc-digital-talent#8671]
## ✨ Feature

We would like update how pool candidate statuses work to make it simpler and compatible with Record of Decision.

## :older_adult:  Old Issues

#3659, #3947

## 🕵️ Details

We will switch our status column and enum to be much simpler: only four states to represent the lifecycle. It will be a "real" column, not derived.
- Draft
- Pending
- Qualified
- Disqualified

We will have only two timestamps to record the movement through those four states.
- submitted_at
- assessed_at

To augment those there will be several timestamps to support scopes and business data.
- expiry_date
- deleted_at
- archived_at
- suspended_at
- removed_at
- placed_at

And several more detail fields:
- assessment_reason (screened out on application, unsuccessful during assessment)
- placement_type (Casual, Term, Indeterminate)
- removal_reason (Requested to be withdrawn, Not responsive, Language testing unsuccessful, Other)
- removal_reason_other (freeform text)

## 🎨 Design File

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/b3450a01-d9b9-49ec-ba95-bd144a7c3c39)

### :car: Migration Plan

| Old Status                  | Status       | Scope             | placement_type | removal_reason       | removal_reason_other                      |
|-----------------------------|--------------|-------------------|----------------|----------------------|-------------------------------------------|
| DRAFT                       | Draft        |                   |                |                      |                                           |
| DRAFT_EXPIRED               | Draft        | Expired           |                |                      |                                           |
| NEW_APPLICATION             | Pending      |                   |                |                      |                                           |
| APPLICATION_REVIEW          | Pending      |                   |                |                      |                                           |
| SCREENED_IN                 | Pending      |                   |                |                      |                                           |
| UNDER_ASSESSMENT            | Pending      |                   |                |                      |                                           |
| SCREENED_OUT_APPLICATION    | Disqualified |                   |                |                      |                                           |
| SCREENED_OUT_NOT_INTERESTED | Pending      | Removed           |                | Withdrawal Requested |                                           |
| SCREENED_OUT_NOT_RESPONSIVE | Pending      | Removed           |                | Not Responsive       |                                           |
| SCREENED_OUT_ASSESSMENT     | Disqualified |                   |                |                      |                                           |
| QUALIFIED_AVAILABLE         | Qualified    |                   |                |                      |                                           |
| QUALIFIED_UNAVAILABLE       | Qualified    | Removed           |                | Not Responsive       |                                           |
| QUALIFIED_WITHDREW          | Qualified    | Removed           |                | Withdrawal Requested |                                           |
| PLACED_CASUAL               | Qualified    | Placed            | Casual         |                      |                                           |
| PLACED_TERM                 | Qualified    | Placed, Suspended | Term           |                      |                                           |
| PLACED_INDETERMINATE        | Qualified    | Placed, Suspended | Indeterminate  |                      |                                           |
| EXPIRED                     | Pending      | Expired           |                |                      |                                           |
| REMOVED                     | Pending      | Removed           |                | Other                | "Migrated from the legacy REMOVED status" |
### :card_file_box: Card Titles
| Status       | Scope                        | placement_type                                                           | TrackApplicationCard       | QualifiedRecruitmentCard                                 |
|--------------|------------------------------|--------------------------------------------------------------------------|----------------------------|----------------------------------------------------------|
| Draft        |                              |                                                                          | "Continue draft"           |                                                          |
| Draft        | Expired                      |                                                                          | "Submission date passed"   |                                                          |
| Pending      |                              |                                                                          | "Application under review" |                                                          |
| Disqualified |                              |                                                                          | "Screened out"             |                                                          |
| Qualified    |                              |                                                                          | "Qualified"                | "Ready for hire"                                         |
| Qualified    | Placed (overrides Suspended) | casual, indeterminate, term                                              |                            | "Hired casual", "Hired indeterminate", "Hired term"      |
| (any)        | Suspended                    |                                                                          | "Not interested"           | "Not interested"                                         |
| (any)        | Removed                      | "request to be withdrawn", "not responsive", "language testing", "other" | "Application on hold"      | "Withdrew", "Paused", "Pending language test", "Removed" |
| (any)        | Expired                      |                                                                          | "Expired"                  | "Expired"                                                |

Refer to https://docs.google.com/spreadsheets/d/16OjYysdlrsSP8IrQJLL7UXYzPA6iskJnlRzeMQ8gDJc/edit#gid=70566376 for the card body and link texts.

### Why is there no placed/hired status?
In a lot of cases we treat someone as placed the same way as qualified, for example when placed casually.  When placed, we will leave users in the Qualified status and automatically set them as suspended when term and indeterminate.  If the user wishes to change the suspended timestamp themselves they can.

## ✅ Acceptance Criteria

- [ ] Database migration  
  - [ ] Map old statuses to new status+detail columns
  - [ ] Rename existing status column to `pool_candidate_status_legacy` to make rollback possible
  - [ ] Status weight generation
  - [ ] New columns (timestamps, details)
- [ ] Ensure new timestamp and detail columns are populated (during mutation?  by model observer?)
- [ ] Update PoolCandidate model scopes
- [ ] Graphql schema update
  - [ ] specialized custom mutations encouraged
- [ ] Unit tests updated
- [ ] frontend update
  - [ ] qualified recruitment cards
  - [ ] track application cards
  - [ ] talent search
  - [ ] applicant status form
  - [ ] define centralized sort order
  - [ ] define centralized colors

## 🛑 Blockers
```[tasklist]
### Blocked By
- [ ] #8415 
- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/issues/10218
```

**Comment by vd1992** [comment]
### Questions:

What splits qualified from screened out? As in the determining field? 

Currently, draft deletion is a hard deletion and I think it was intentionally set as such. Is there a reason to soft delete them instead?

Would this status then be a virtual property obtained from an accessor on the pool candidate model? Computed from the various fields that returns an enum?

### Notes:

There are multiple options for timestamps updates. Like:

Submitted from the client

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/ef511631-807a-472d-bb97-02e5140c0418)

The injectNow directive

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/c4f96412-2c20-435d-96f0-3fb9e47b929f)

An setter/mutator that exists to tack on a model change. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/e0a5c62f-b387-463c-91bd-1158dcb2fe68)

**Comment by petertgiles** [comment]
If we added a field for just the assessment decision, like "SUCCESSFUL" or "UNSUCCESSFUL" we could get away without a "real" status column.  We would still need a real status detail column though.

**Comment by vd1992** [comment]
### Fields

**Existing fields are:**
submitted_at
expiry_date (not a timestamp, will this pose trouble?)
archived_at
suspended_at
deleted_at

**To add timestamps:**
removed_at
assessed_at
placed_at

**Strings using enums:**
status, 4 options 
assessment_reason, two options
placement_type, three options
removal_reason, four options

**Text fields to add:**
removal_reason_other

### Other 

Status override/derived field to be removed
Migration when setting scope fields, just inject the current time I imagine

### Questions

How will filtering by status now work? The scopes can be written up simply enough but how will the Client-API connection look? What will users see? 

Where will status be set automatically and where will it be controlled? 
For instance, draft is upon creation, pending in the submit action mutation, but where will the switch to qualified/unqualified happen? 

Is creating new mutations for the new fields and where they will live in scope? Like the removed field, updating it.

**Comment by tristan-orourke** [comment]
We've decided to hold off on this and try to launch Record of Decision with the current implementation for application statuses, to reduce the complexity of this release. We will map the current implementation of statuses to the new UI in the frontend. Later, we can come back to this and refactor the status implementation.

**Comment by gobyrne** [comment]
@tristan-orourke is this still something we intend to do?

**Comment by esizer** [comment]
We have had some tickets related to statuses recently (to derive it from dates/final decision). Is this still relevant with that work happening?
